doc.go

```diff
@@ -0,0 +1,175 @@
+instead, their data array contains links to child nodes. The index compaction
+using the bitmap is the same as for leaf nodes, so each non-leaf node only
+stores as many links as it has child nodes.
+
+Because additional height is required to address larger indexes, even a
+single-element AMT will require more than one node where the index is greater
+than the width of the AMT. For a width of 8, indexes 8 to 63 require a height
+of 1, indexes 64 to 511 require a height of 2, indexes 512 to 4095 require a
+height of 3, etc.
+
+Retrieving elements from the AMT requires extracting only the portion of the
+requested index that is required at each height to determine the position in
+the data array to navigate into. When traversing through the tree, we only need
+to select from indexes 0 to width-1. To do this, we take log2(width) bits from
+the index to form a number that is between 0 and width-1. e.g. for a width of
+8, we only need 3 bits to form a number between 0 and 7, so we only consume
+3 bits per level of the AMT as we traverse. A simple method to calculate this
+at any height in the AMT (assuming bitWidth of 3, i.e. a width of 8) is:
+
+1. Calculate the maximum number of nodes (not entries) that may be present in
+an sub-tree rooted at the current height. width^height provides this number.
+e.g. at height 0, only 1 node can be present, but at height 3, we may have a
+tree of up to 512 nodes (storing up to 8^(3+1)=4096 entries).
+
+2. Divide the index by this number to find the index for this height. e.g. an
+index of 3 at height 0 will be 3/1=3 , or an index of 20 at height will be
+20/8=2.
+
+3. If we are at height 0, the element we want is at the data index,
+position-mapped via the bitmap.
+
+4. If we are above height 0, we need to navigate to the child element at the
+index we calculated, position-mapped via the bitmap. When traversing to the
+child, we discard the upper portion of the index that we no longer need.
+This can be achieved by a mod operation against the number-of-nodes value.
+e.g. an index of 20 at height 1 requires navigation to the element at
+position 2, when moving to that element (which is height 0), we truncate the
+index with 20%8=4, at height 0 this index will be the index in our data
+array (position-mapped via the bitmap).
+
+In this way, each sub-tree root consumes a small slice, log2(width) bits long,
+of the original index.
+
+Adding new elements to an AMT may require up to 3 steps:
+
+1. Increasing the height to accommodate a new index if the current height is not
+sufficient to address the new index. Increasing the height requires turning
+the current root node into an intermediate and adding a new root which
+links to the old (repeated until the required height is reached).
+
+2. Adding any missing intermediate and leaf nodes that are required at the
+index required. There may be large gaps in the AMT which will require
+filling with new nodes.
```

Comment:
This is a bit too fuzzy to me. Isn't it more concretely (when adding a single new node) adding intermediate nodes up to the top of the root? So it would be `height - 1` (where height is the tree height) new intermediate nodes at most?

Reply:
I've expanded a bit on these steps to make it clearer, it's all about filling in the required height to address the new indexes which may not already have the connecting intermediates.